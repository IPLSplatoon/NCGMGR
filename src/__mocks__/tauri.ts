export const mockTauriEvent = {
    listen: jest.fn()
}

jest.mock('@tauri-apps/api/event', () => mockTauriEvent)

export const mockTauri = {
    invoke: jest.fn()
}

jest.mock('@tauri-apps/api/tauri', () => mockTauri)

export const mockTauriDialog = {
    open: jest.fn()
}

jest.mock('@tauri-apps/api/dialog', () => mockTauriDialog)

export const mockTauriFs = {
    readDir: jest.fn(),
    readTextFile: jest.fn(),
    removeDir: jest.fn(),
    removeFile: jest.fn(),
    createDir: jest.fn(),
    writeFile: jest.fn()
}

jest.mock('@tauri-apps/api/fs', () => mockTauriFs)

export const mockTauriOs = {
    type: jest.fn()
}

jest.mock('@tauri-apps/api/os', () => mockTauriOs)

export const mockTauriShell = {
    open: jest.fn()
}

jest.mock('@tauri-apps/api/shell', () => mockTauriShell)
