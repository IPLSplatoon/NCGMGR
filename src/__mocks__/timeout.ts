export function mockTimeout (): void {
    jest.spyOn(window, 'setTimeout').mockImplementation(handler => {
        // noinspection SuspiciousTypeOfGuard
        if (typeof handler !== 'string') {
            handler()
        }
        return 0
    })
}
