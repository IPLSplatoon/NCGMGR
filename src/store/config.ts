import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const configStore = createStore<Configuration>({
    state: {
        installPath: ''
    },
    mutations: {
        setInstallPath: (state, installPath: string) => {
            state.installPath = installPath
        },
        setState: (state, config: Configuration) => {
            Object.assign(state, config)
        }
    },
    actions: {
        load: store => {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            store.commit('setState', {
                ...store.state,
                ...config
            })
        },
        persist: store => {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(store.state))
        }
    }
})

export const configStoreKey: InjectionKey<Store<Configuration>> = Symbol()

export function useConfigStore (): Store<Configuration> {
    return useStore(configStoreKey)
}
