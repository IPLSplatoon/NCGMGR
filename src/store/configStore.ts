import { defineStore } from 'pinia'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const useConfigStore = defineStore('config', {
    state: () => ({
        installPath: ''
    } as Configuration),
    actions: {
        load () {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            this.$patch(config)
        },
        persist () {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(this.$state))
        }
    }
})
