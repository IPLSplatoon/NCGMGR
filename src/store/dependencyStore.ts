import { defineStore } from 'pinia'
import { isBlank } from '@iplsplatoon/vue-components'
import { getNodejsVersion } from '@/service/dependencyService'

interface DependencyStore {
    isLoading: boolean
    nodejs: {
        version: string | null
    }
}

export const useDependencyStore = defineStore('dependency', {
    state: () => ({
        isLoading: false,
        nodejs: {
            version: null
        }
    } as DependencyStore),
    getters: {
        hasNodejs: state => !isBlank(state.nodejs.version)
    },
    actions: {
        async checkVersions () {
            this.isLoading = true
            try {
                this.nodejs.version = await getNodejsVersion()
            } finally {
                this.isLoading = false
            }
        }
    }
})
