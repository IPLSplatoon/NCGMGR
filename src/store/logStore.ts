import { ActionState, LogEvent, ProgressEvent } from '@/types/log'
import { listen, UnlistenFn } from '@tauri-apps/api/event'
import { defineStore } from 'pinia'

export interface LogStore {
    lines: Record<string, LogEvent[]>
    progressEntries: Record<string, ProgressEvent>
    actionStates: Record<string, ActionState>
}

const unlistenFns: Record<string, UnlistenFn[]> = {}

export const useLogStore = defineStore('log', {
    state: () => ({
        lines: {},
        progressEntries: {},
        actionStates: {}
    } as LogStore),
    actions: {
        insertLine ({ line, key }: { line: LogEvent, key: string }) {
            if (!this.lines[key]) {
                this.lines[key] = []
            }
            this.lines[key].push(line)
        },
        setProgress (key: string, event: ProgressEvent) {
            const existingEntry = this.progressEntries[key]
            if (existingEntry == null) {
                this.progressEntries[key] = event
                this.logProgressEvent(key, event)
            } else {
                existingEntry.message = event.message
                if (event.maxStep != null) {
                    existingEntry.maxStep = event.maxStep
                }
                if (event.step != null) {
                    existingEntry.step = event.step

                    if (event.step === event.maxStep) {
                        this.setActionState({ key, state: ActionState.COMPLETED_SUCCESS })
                    }
                }

                this.logProgressEvent(key, {
                    message: event.message,
                    step: event.step,
                    maxStep: existingEntry.maxStep
                })
            }
        },
        logProgressEvent(key: string, event: ProgressEvent) {
            if (event.step != null) {
                this.insertLine({
                    line: {
                        message: `[${event.step}/${event.maxStep}] ${event.message}`,
                        type: 'info'
                    },
                    key
                })
            } else {
                this.insertLine({
                    line: {
                        message: event.message,
                        type: 'info'
                    },
                    key
                })
            }
        },
        reset (key: string) {
            delete this.progressEntries[key]
            this.lines[key] = []
            this.actionStates[key] = ActionState.INCOMPLETE
        },
        setActionState ({ key, state }: { key: string, state: ActionState }) {
            this.actionStates[key] = state
        },
        listen (key: string, listenForProgress = false) {
            if (unlistenFns[key]) {
                return
            }

            const listenPromises: Array<Promise<UnlistenFn>> = []

            listenPromises.push(listen<LogEvent>(`log:${key}`, event => {
                this.insertLine({ line: event.payload, key })
            }))

            if (listenForProgress) {
                listenPromises.push(listen<ProgressEvent>(`progress:${key}`, event => {
                    this.setProgress(key, event.payload)
                }))
            }

            return Promise.all(listenPromises).then(result => {
                unlistenFns[key] = result
            })
        },
        unlisten (key: string) {
            if (unlistenFns[key]) {
                unlistenFns[key].forEach(fn => fn())
                delete unlistenFns[key]
            }
        },
        logPromiseResult ({ promise, key }: { promise: Promise<unknown>, key: string }) {
            promise.catch(e => {
                this.insertLine({
                    line: {
                        message: String(e),
                        type: 'error'
                    },
                    key
                })
                this.setActionState({ key, state: ActionState.COMPLETED_ERROR })
            })
        }
    }
})
