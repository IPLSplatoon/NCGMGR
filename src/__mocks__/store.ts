import { createStore, Store } from 'vuex'
import { Configuration } from '@/store/config'
import { LogStore } from '@/store/log'

export function createConfigStore (): Store<Configuration> {
    return createStore<Configuration>({
        state: {
            installPath: '/install/path'
        },
        actions: {
            load: jest.fn(),
            persist: jest.fn()
        },
        mutations: {
            setInstallPath: jest.fn()
        }
    })
}

export function createLogStore (): Store<LogStore> {
    return createStore<LogStore>({
        state: {
            lines: [],
            completed: false
        },
        mutations: {
            reset: jest.fn(),
            insertLine: jest.fn(),
            setCompleted: jest.fn()
        }
    })
}
