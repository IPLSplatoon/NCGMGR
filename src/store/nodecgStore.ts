import { Bundle, getBundles, getNodecgStatus } from '@/service/nodecgService'
import { useConfigStore } from '@/store/configStore'
import { defineStore } from 'pinia'
import { listenForProcessExit } from '@/service/messagingService'

export enum InstallStatus {
    UNKNOWN,
    READY_TO_INSTALL,
    INSTALLED,
    UNABLE_TO_INSTALL,
    MISSING_INSTALL_DIRECTORY,
    BAD_INSTALL_DIRECTORY
}

export enum RunStatus {
    NOT_STARTED = 'NOT_STARTED',
    RUNNING = 'RUNNING',
    STOPPED = 'STOPPED'
}

export interface NodecgStore {
    status: {
        installStatus: InstallStatus
        runStatus: RunStatus
        message: string
        bundlesLoading: boolean
    },
    bundles: Bundle[]
}

export const useNodecgStore = defineStore('nodecg', {
    state: () => ({
        status: {
            installStatus: InstallStatus.UNKNOWN,
            runStatus: RunStatus.NOT_STARTED,
            message: '',
            bundlesLoading: false
        },
        bundles: []
    } as NodecgStore),
    actions: {
        async checkNodecgStatus () {
            const configStore = useConfigStore()

            try {
                const { status, message } = await getNodecgStatus(configStore.userConfig.nodecgInstallDir)
                this.status.message = message
                this.status.installStatus = status
                if (status === InstallStatus.INSTALLED) {
                    this.getBundleList()
                } else {
                    this.bundles = []
                }
            } catch (e) {
                this.status.message = String(e)
                this.status.installStatus = InstallStatus.UNABLE_TO_INSTALL
                this.bundles = []
            }
        },
        async getBundleList () {
            const configStore = useConfigStore()
            this.status.bundlesLoading = true
            try {
                this.bundles = await getBundles(configStore.userConfig.nodecgInstallDir)
            } finally {
                this.status.bundlesLoading = false
            }
        }
    }
})

listenForProcessExit('run-nodecg', () => {
    useNodecgStore().$state.status.runStatus = RunStatus.STOPPED
})
