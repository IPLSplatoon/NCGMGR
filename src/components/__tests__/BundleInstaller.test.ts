import { mockTauri } from '@/__mocks__/tauri'
import BundleInstaller from '@/components/BundleInstaller.vue'
import { config, flushPromises, mount } from '@vue/test-utils'
import Mock = jest.Mock
import { IplButton, IplInput } from '@iplsplatoon/vue-components'
import { createTestingPinia, TestingPinia } from '@pinia/testing'
import { useConfigStore } from '@/store/configStore'
import { useLogStore } from '@/store/logStore'
import { useNodecgStore } from '@/store/nodecgStore'

describe('BundleInstaller', () => {
    config.global.stubs = {
        IplInput: true,
        IplButton: true,
        LogOverlay: true
    }

    let pinia: TestingPinia

    beforeEach(() => {
        pinia = createTestingPinia()
        config.global.plugins = [pinia]
    })

    it('matches snapshot', () => {
        const wrapper = mount(BundleInstaller)

        expect(wrapper.html()).toMatchSnapshot()
    })

    it('handles installation', async () => {
        const logStore = useLogStore()
        logStore.reset = jest.fn()
        logStore.logPromiseResult = jest.fn()
        logStore.listenForProcessExit = jest.fn()
        logStore.listen = jest.fn()
        const configStore = useConfigStore()
        configStore.installPath = '/install/path'
        const nodecgStore = useNodecgStore()
        nodecgStore.getBundleList = jest.fn()
        mockTauri.invoke.mockResolvedValue({})
        const wrapper = mount(BundleInstaller, {
            global: {
                plugins: [pinia]
            }
        })

        wrapper.getComponent<typeof IplInput>('[data-test="bundle-path-input"]').vm.$emit('update:modelValue', 'git://new-path')
        await flushPromises()
        wrapper.getComponent<typeof IplButton>('[data-test="install-button"]').vm.$emit('click')
        await flushPromises()

        expect(logStore.reset).toHaveBeenCalledWith('install-bundle')
        expect(logStore.listen).toHaveBeenCalledWith('install-bundle', true)
        expect(mockTauri.invoke).toHaveBeenCalledWith('install_bundle', { bundleUrl: 'git://new-path', nodecgPath: '/install/path' })
        expect(logStore.logPromiseResult).toHaveBeenCalledWith({ promise: expect.anything(), key: 'install-bundle' })
        expect(logStore.listenForProcessExit).toHaveBeenCalledWith({ callback: expect.anything(), key: 'install-bundle' });
        (logStore.listenForProcessExit as Mock).mock.calls[0][0].callback()
        expect(nodecgStore.getBundleList).toHaveBeenCalled()
        expect(wrapper.findComponent('[data-test="bundle-log-overlay"]').exists()).toEqual(true)
    })
})
