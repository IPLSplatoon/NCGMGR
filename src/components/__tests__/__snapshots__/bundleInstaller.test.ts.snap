// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BundleInstaller matches snapshot 1`] = `
<div class="ipl-space color-light">
  <ipl-input-stub name="bundleName" label="Bundle Repository Path" modelvalue="" type="text" validator="[object Object]" centered="false" disabled="false" data-test="bundle-path-input"></ipl-input-stub>
  <ipl-button-stub label="Install" color="green" disabled="true" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" nobackground="false" tooltip="" class="m-t-8" data-test="install-button"></ipl-button-stub>
  <log-overlay-stub title="Installing..." visible="false" data-test="bundle-log-overlay"></log-overlay-stub>
</div>
`;
